### PR TITLE
Chat UI tweaks

### PR DESCRIPTION
**Pull request by iojw**
Simple fix for #2376 

### Contains

- Chat overlay now shows up to last 6 messages
- Sending a message now closes the chat window and opens the chat overlay instead of remaining focused on the text field
- Tweaked base time values of chat overlay so it lasts longer before fading out

### How to test

Open up a game and type something in chat. Here's an image of the updated chat overlay:

### Outstanding before merging

Some ideas I had but not really sure if they should be put in the game:
- Settings for max no. of lines to be shown on overlay
- Similar to other games, where the chat history is separate from the text box and pressing the chat button just makes the text field visible and turns the chat history scrollable (my concern with this is that it might take up too much space since the new messages appear at the bottom but we could change this such that new messages appear on top)

I adjusted the chat system based on my own judgement and experience with other games so definitely let me know if you would like any changes and I can add them in 🙂.

### CODE REVIEW

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

File: engine/src/main/java/org/terasology/logic/console/ui/MiniChatOverlay.java [hunk 1]
```diff
@@ -62,13 +65,21 @@ public void initialise() {
...
 
-                return last;
+                //checks if number of messages > MAX_LINES, else starts from 0
+                for (int i = msgsList.size() >= MAX_LINES ? msgsList.size() - MAX_LINES : 0; i < msgsList.size(); i++) {
+                    messageHistory.append(msgsList.get(i).getMessage());
+                    if (i < msgsList.size() - 1) {
+                        messageHistory.append(Console.NEW_LINE);
+                    }
+                }
+                return messageHistory.toString();
```
Comment (oniatus) [inline engine/src/main/java/org/terasology/logic/console/ui/MiniChatOverlay.java]: Not sure what is on the classpath, but this looks like a good use for [EvictingQueue](http://google.github.io/guava/releases/snapshot/api/docs/com/google/common/collect/EvictingQueue.html) or [CircularFifoQueue](http://commons.apache.org/proper/commons-collections/javadocs/api-release/org/apache/commons/collections4/queue/CircularFifoQueue.html).
Or even better: Build the string while iterating over the messageList, if you use the counter anyway.

Comment (iojw) [inline engine/src/main/java/org/terasology/logic/console/ui/MiniChatOverlay.java]: What do you mean by "Build the string while iterating over the messageList"? Do you mean using Iterables.size() and a counter to append the last 6 messages?

Comment (oniatus) [inline engine/src/main/java/org/terasology/logic/console/ui/MiniChatOverlay.java]: That sounds preferable to copying the entire buffer first and pruning the copy to the last n messages afterwards.

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

File: engine/src/main/java/org/terasology/logic/console/ui/MiniChatOverlay.java (hunk 1 above)
Comment (oniatus) [inline engine/src/main/java/org/terasology/logic/console/ui/MiniChatOverlay.java]: Checked out and worked well so far.
Small experiment on readability (not sure what to prefer, I would leave this up to you):
```
        message.bindText(new ReadOnlyBinding<String>() {
            @Override
            public String get() {
                Iterable<Message> messages = console.getMessages(CoreMessageType.CHAT, CoreMessageType.NOTIFICATION);
                int messagesCount = Iterables.size(messages);
                int messagesToSkip = messagesCount - MAX_LINES;
                if (messagesToSkip > 0) {
                    messages = Iterables.skip(messages, messagesToSkip);
                }

                Stream<Message> mostRecentMessages = StreamSupport.stream(messages.spliterator(), false);
                String messagesHistory = mostRecentMessages.map(Message::getMessage).collect(Collectors.joining(Console.NEW_LINE));
                return messagesHistory;
            }
        });
```

In any case, `MAX_LINES` should probably be `MAX_MESSAGES` or `MAX_VISIBLE_MESSAGES`. Fun fact: The overlay can get really huge, when large messages wrap over multiple lines (tested with large `BANANA...AS` 👓 . However adding a scrollpane or a real line limitiation/char is likely out of scope for this PR, so this is fine from my side :)

**Comment by oniatus** [comment]
Whoops, put my [last comment](https://github.com/MovingBlocks/Terasology/pull/2667/files/b447ec38829b0710b9fceb3b1464a4c0f7a6ead2#r92462459) in the first commit :)

Just give me a short ping if you want to do some more stuff on the code, otherwhise I think we can merge this one tomorrow :)

**Comment by iojw** [comment]
@oniatus: Ah, didn't think about that. Would like to deal with it before it gets merged. I don't think we should add a scrollbar since that's already present in the full chat history or a char limit since that might lead to one message filling the whole overlay.. It does seem strange that there's no limit for chat messages - what do you about limiting all messages sent to 100 chars? 🙂

**Comment by oniatus** [comment]
@iojw I think the overlay has to deal with the message size, not vice versa :) If one wants to copy-paste a novel or a large link, why not.
But as the issue was about improving the visibility and adding a fade out, this is okay for the moment.
We can add a follow-up issue about a maximum size or more polishing like different or configurable fade out times if you want to continue on this 👍

**Comment by portokaliu** [comment]
The only thing I can think about regarding this is having a fade to a larger text area (with previously buffered messages) that can only be seen when trying to type stuff. Probably, as you said, out of the current scope though

**Comment by iojw** [comment]
@oniatus Cool, I guess you can just merge this in then 🙂

@portokaliu That should be what what happens currently - this overlay is only shown when the player is not actively typing and someone else sends a message. When the player is actively typing, there's a larger scrollable view visible with the full chat history.

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Pull request merged**